src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreePostingsWriter.java

```diff
@@ -0,0 +1,198 @@
+     * level 0, because there is no node at level 0).
+     */
+    public static final int POSTINGS_SKIP = CassandraRelevantProperties.SAI_POSTINGS_SKIP.getInt();
+
+    private final List<PackedLongValues> postings;
+    private final TreeMap<Long, Integer> leafOffsetToNodeID = new TreeMap<>(Long::compareTo);
+    private final Multimap<Integer, Integer> nodeToChildLeaves = HashMultimap.create();
+
+    private final IndexContext indexContext;
+    int numNonLeafPostings = 0;
+    int numLeafPostings = 0;
+
+    BlockBalancedTreePostingsWriter(List<PackedLongValues> postings, IndexContext indexContext)
+    {
+        this.postings = postings;
+        this.indexContext = indexContext;
+    }
+
+    @Override
+    public void onLeaf(int leafNodeID, long leafBlockFP, IntArrayList pathToRoot)
+    {
+        checkArgument(!pathToRoot.containsInt(leafNodeID));
+        checkArgument(pathToRoot.isEmpty() || leafNodeID > pathToRoot.get(pathToRoot.size() - 1));
+
+        leafOffsetToNodeID.put(leafBlockFP, leafNodeID);
+        for (int i = 0; i < pathToRoot.size(); i++)
+        {
+            int level = i + 1;
+            if (isLevelEligibleForPostingList(level))
+            {
+                int nodeID = pathToRoot.get(i);
+                nodeToChildLeaves.put(nodeID, leafNodeID);
+            }
+        }
+    }
+
+    public long finish(IndexOutput out) throws IOException
+    {
+        checkState(postings.size() == leafOffsetToNodeID.size(),
+                   "Expected equal number of postings lists (%s) and leaf offsets (%s).",
+                   postings.size(), leafOffsetToNodeID.size());
+
+        try (PostingsWriter postingsWriter = new PostingsWriter(out))
+        {
+            Iterator<PackedLongValues> postingsIterator = postings.iterator();
+            Map<Integer, PackedLongValues> leafToPostings = new HashMap<>();
+            leafOffsetToNodeID.forEach((fp, nodeID) -> leafToPostings.put(nodeID, postingsIterator.next()));
+
+            long postingsRamBytesUsed = postings.stream()
+                                                .mapToLong(PackedLongValues::ramBytesUsed)
+                                                .sum();
+
+            List<Integer> internalNodeIDs = nodeToChildLeaves.keySet()
+                                                             .stream()
+                                                             .filter(i -> nodeToChildLeaves.get(i).size() >= MINIMUM_POSTINGS_LEAVES)
+                                                             .collect(Collectors.toList());
+
+            Collection<Integer> leafNodeIDs = leafOffsetToNodeID.values();
+
+            logger.debug(indexContext.logMessage("Writing posting lists for {} internal and {} leaf balanced tree nodes. Leaf postings memory usage: {}."),
+                         internalNodeIDs.size(),
+                         leafNodeIDs.size(),
+                         FBUtilities.prettyPrintMemory(postingsRamBytesUsed));
```

Comment:
nit: Honestly would be fine w/ these 3 arguments on one line. Helps me line them up w/ the `{}`s above.